# Data Analysis (part 1)
## Importing, Cleaning, and Exploring COMM data 

In [1]:
%matplotlib inline

import json
import pandas as pd
import matplotlib.pyplot as plt
from pandas.io.json import json_normalize
import numpy as np
import matplotlib.pyplot as plt
import requests
import warnings
warnings.filterwarnings('ignore')
import seaborn as sns
import plotly_express as px
import math
import plotly

#### Importing COMM courses for analysis
* First we need to use the Penn Course Review API and the Python requests package to import a JSON file 
* The non-public API key was granted after application via Penn Labs, the student-run club that manages PCR. 

In [2]:
comm_revs1 = requests.get('http://api.penncoursereview.com/v1/depts/comm/reviews/?token=rPY7nzxkE0Dqjzlu9LUDX2mV30W0qo').json()

In [3]:
comm_revs1

{'result': {'values': [{'comments': '',
    'id': '292-001-277-JESSICA-FISHMAN',
    'instructor': {'first_name': 'JESSICA',
     'id': '277-JESSICA-FISHMAN',
     'last_name': 'FISHMAN',
     'name': 'JESSICA FISHMAN',
     'path': '/instructor/277-JESSICA-FISHMAN'},
    'num_reviewers': 83,
    'num_students': 89,
    'path': '/courses/292/sections/1/reviews/277-JESSICA-FISHMAN',
    'ratings': {'rAmountLearned': '2.63',
     'rCommAbility': '2.88',
     'rCourseQuality': '2.43',
     'rDifficulty': '2.28',
     'rInstructorAccess': '2.84',
     'rInstructorQuality': '2.88',
     'rReadingsValue': '2.24',
     'rRecommendMajor': '3.21',
     'rRecommendNonMajor': '2.53',
     'rStimulateInterest': '2.81',
     'rWorkRequired': '2.28'},
    'section': {'aliases': ['COMM-123-001'],
     'id': '292-001',
     'name': 'COMM & POPULAR CULTURE',
     'path': '/courses/292/sections/1',
     'primary_alias': 'COMM-123-001',
     'sectionnum': '001',
     'semester': '2002A'}},
   {'comments'

#### This is a layered JSON file. We need to normalize and parse the data:

In [4]:
comm_revs_df = json_normalize(comm_revs1['result']['values'])

Now let's take a look at the rows and columns.

In [5]:
comm_revs_df

comments                                            id  \
0                              292-001-277-JESSICA-FISHMAN   
1                                293-001-278-VINCENT-PRICE   
2                        293-601-279-MARIAELENA-BARTESAGHI   
3                              294-401-280-KATHRYN-KOLBERT   
4                                295-001-281-AMY-B--JORDAN   
5                                   296-601-282-EMILY-WEST   
6                           297-601-283-JOSEPHINE-FERRIGNO   
7                            298-001-284-RONA-J--BUCHALTER   
8                             299-301-285-CAROLYN-A-MARVIN   
9                    299-601-286-MARGARET-LOUISE-WOODSTOCK   
10                             300-001-287-JOSEPH-G--TUROW   
11                          301-301-288-OSCAR-HOWARD-GANDY   
12                          302-001-289-JOSEPH-N--CAPPELLA   
13                          303-301-290-KLAUS-KRIPPENDORFF   
14                        304-301-291-DWIGHT-D--EISENHOWER   
15                                305-302-292-JOHN-JEMMOTT   
16                               305-303-293-MICHAEL-HAGEN   
17                                  306-301-294-ELIHU-KATZ   
18                                 307-301-295-ALBERT-HUNT   
19                          1453-601-284-RONA-J--BUCHALTER   
20                            1454-001-287-JOSEPH-G--TUROW   
21                           1455-001-1813-CARRIE-D-RICKEY   
22                     1456-001-1814-DEBORAH-L--LINEBARGER   
23                          1456-601-1815-KIMBERLY-D-WOOLF   
24                             1457-601-1336-CLAIRE-WARDLE   
25                        1458-001-1816-AMY-SARCH-SCHOPICK   
26                         1459-001-289-JOSEPH-N--CAPPELLA   
27                       1460-001-1099-KRYSTYNA-B--WARCHOL   
28                                1461-301-295-ALBERT-HUNT   
29                       1462-301-291-DWIGHT-D--EISENHOWER   
..       ...                                           ...   
730                       47088-001-289-JOSEPH-N--CAPPELLA   
731                            47089-301-8870-LIK-SAM-CHAN   
732                     47090-001-7199-DWAYNE-TAYLOR-BOOTH   
733                 47091-301-8374-MATTHEW-BROOK-O-DONNELL   
734                        47092-301-8650-AARON-M--SHAPIRO   
735                         47093-301-285-CAROLYN-A-MARVIN   
736                              47094-301-295-ALBERT-HUNT   
737                   47095-301-8871-RAYYA-SUNAYMA-EL-ZEIN   
738                     47096-301-8872-DANIEL-RAY-GRINBERG   
739                        47097-301-2904-CARLIN-P--ROMANO   
740                            47098-301-8873-EYTAN-GILBOA   
741                     47099-301-291-DWIGHT-D--EISENHOWER   
742                       47099-301-7990-SCOTT-DAVID-REICH   
743                          47100-301-5327-KENNETH-WINNEG   
744                  47101-301-7003-SANDRA-GONZALEZ-BAILON   
745                          47102-301-8229-YPHTACH-LELKES   
746                        47103-301-8874-DAMON-M--CENTOLA   
747                           47104-301-6859-SUSAN-D--HAAS   
748                          47923-401-7291-DAVID-J-DUTWIN   
749                           47987-401-8648-MURALI-BALAJI   
750                             48007-401-6592-GUOBIN-YANG   
751                    48197-910-9018-MARIA-CELESTE-WAGNER   
752                          48197-910-9019-PAWEL-M-POPIEL   
753                    48198-920-9020-LAUREN-EMILY-BRIDGES   
754                        48198-920-9021-DIAMI-J-VIRGILIO   
755                    48199-910-9022-LEEANN-NICOLE-SIEGEL   
756                48200-900-9023-MICHAEL-RYAN-TSAPATSARIS   
757           48200-910-9024-JASMINE-ERIN-NORKUNAS-ERDENER   
758                          48200-910-9025-HANNA-E-MORRIS   
759                 48201-920-9026-JENNIFER-RAE-HENRICHSEN   

     instructor.first_name                       instructor.id  \
0                  JESSICA                 277-JESSICA-FISHMAN   
1                  VINCENT 

Each row/observation seems to be one course, and the columns are units of analysis featuring different variables from student-submitted course evaluations.

In [6]:
# saving this to the raw data folder
comm_revs_df.to_csv('../data/raw_data/comm_revs_df',index=False)


#### Inspecting the columns

In [7]:
comm_revs_df.columns

Index(['comments', 'id', 'instructor.first_name', 'instructor.id',
       'instructor.last_name', 'instructor.name', 'instructor.path',
       'num_reviewers', 'num_students', 'path', 'ratings.rAmountLearned',
       'ratings.rCommAbility', 'ratings.rCourseQuality', 'ratings.rDifficulty',
       'ratings.rInstructorAccess', 'ratings.rInstructorQuality',
       'ratings.rReadingsValue', 'ratings.rRecommendMajor',
       'ratings.rRecommendNonMajor', 'ratings.rStimulateInterest',
       'ratings.rTAQuality', 'ratings.rWorkRequired', 'section.aliases',
       'section.id', 'section.name', 'section.path', 'section.primary_alias',
       'section.sectionnum', 'section.semester'],
      dtype='object')

#### This dataset needs some cleaning. Let's begin with the columns.
* There are some columns we simply do not need for this analysis.
* Others we will want to rename
* We will need the columns with ratings to be treated as numeric so we can do statistical analyses
* We'll add a column just called "level" to represent the course code
* And we will want to re-order the columns to be more user-friendly

In [8]:
# dropping some columns we do not need
comm_revs_df2 = comm_revs_df.drop(columns=['comments','id', 'instructor.id', 'instructor.name','instructor.path','path','section.aliases',
                          'section.id','section.path',])

In [9]:
# sanity check to make sure that they actually dropped cleanly
comm_revs_df2.columns

Index(['instructor.first_name', 'instructor.last_name', 'num_reviewers',
       'num_students', 'ratings.rAmountLearned', 'ratings.rCommAbility',
       'ratings.rCourseQuality', 'ratings.rDifficulty',
       'ratings.rInstructorAccess', 'ratings.rInstructorQuality',
       'ratings.rReadingsValue', 'ratings.rRecommendMajor',
       'ratings.rRecommendNonMajor', 'ratings.rStimulateInterest',
       'ratings.rTAQuality', 'ratings.rWorkRequired', 'section.name',
       'section.primary_alias', 'section.sectionnum', 'section.semester'],
      dtype='object')

We should inspect the dataframe now to see that we are getting closer to something we can work with.

In [10]:
comm_revs_df2.head(5)

instructor.first_name instructor.last_name  num_reviewers  num_students  \
0               JESSICA              FISHMAN             83            89   
1               VINCENT                PRICE            133           181   
2            MARIAELENA           BARTESAGHI             23            26   
3               KATHRYN              KOLBERT             73           112   
4                AMY B.               JORDAN             75            89   

  ratings.rAmountLearned ratings.rCommAbility ratings.rCourseQuality  \
0                   2.63                 2.88                   2.43   
1                   2.33                 2.84                   2.11   
2                   2.65                 3.17                   2.74   
3                   1.68                 2.37                   1.89   
4                   3.08                 3.49                   3.20   

  ratings.rDifficulty ratings.rInstructorAccess ratings.rInstructorQuality  \
0                2.28                      2.84                       2.88   
1                2.27                      2.62                       2.62   
2                2.13                      3.61                       3.17   
3                1.49                      2.34                       2.26   
4                2.13                      3.24                       3.49   

  ratings.rReadingsValue ratings.rRecommendMajor ratings.rRecommendNonMajor  \
0                   2.24                    3.21                       2.53   
1                   2.14                    3.15                       1.47   
2                   3.17                    3.17                       2.09   
3                   1.99                    2.42                       1.86   
4                   2.49                    3.60                       2.87   

  ratings.rStimulateInterest ratings.rTAQuality ratings.rWorkRequired  \
0                       2.81                NaN                  2.28   
1                       1.92                NaN                  2.32   
2                       3.30                NaN                  2.13   
3                       1.79                NaN                  1.83   
4                       3.07                NaN                  2.29   

               section.name section.primary_alias section.sectionnum  \
0    COMM & POPULAR CULTURE          COMM-123-001                001   
1    COMMUNICATION BEHAVIOR          COMM-125-001                001   
2    COMMUNICATION BEHAVIOR          COMM-125-601                601   
3  ARGUMENT & PUBL ADVOCACY          COMM-175-401                401   
4          CHILDREN & MEDIA          COMM-225-001                001   

  section.semester  
0            2002A  
1            2002A  
2            2002A  
3            2002A  
4            2002A

These column names are not so user-friendly. Making them more familiar and useful now:

In [11]:
#renaming columns
comm_revs_df2.columns = ['instructor_first', 'instructor_last', 
                         'num_reviewers','num_students', 
                      'AmountLearned',
       'CommAbility', 'CourseQuality', 'Difficulty',
       'InstructorAccess', 'InstructorQuality',
       'ReadingsValue', 'RecommendMajor',
       'RecommendNonMajor', 'StimulateInterest',
       'TAQuality', 'WorkRequired', 'CourseTitle',
       'CourseCode', 'Section', 'Semester']

Checking out the new dataframe.

In [12]:
comm_revs_df2

instructor_first  instructor_last  num_reviewers  num_students  \
0                  JESSICA          FISHMAN             83            89   
1                  VINCENT            PRICE            133           181   
2               MARIAELENA       BARTESAGHI             23            26   
3                  KATHRYN          KOLBERT             73           112   
4                   AMY B.           JORDAN             75            89   
5                    EMILY             WEST             23            24   
6                JOSEPHINE         FERRIGNO             18            25   
7                  RONA J.        BUCHALTER             13            14   
8                CAROLYN A           MARVIN             17            18   
9          MARGARET LOUISE        WOODSTOCK             17            18   
10               JOSEPH G.            TUROW             56            92   
11            OSCAR HOWARD            GANDY              5             6   
12               JOSEPH N.         CAPPELLA             48            56   
13                   KLAUS     KRIPPENDORFF             15            19   
14               DWIGHT D.       EISENHOWER             22            23   
15                    JOHN          JEMMOTT              6             6   
16                 MICHAEL            HAGEN             12            16   
17                   ELIHU             KATZ              5            21   
18                  ALBERT             HUNT             14            31   
19                 RONA J.        BUCHALTER             24            26   
20               JOSEPH G.            TUROW             94           192   
21                CARRIE D           RICKEY             49            79   
22              DEBORAH L.       LINEBARGER             26            37   
23              KIMBERLY D            WOOLF             21            24   
24                  CLAIRE           WARDLE             18            20   
25               AMY SARCH         SCHOPICK             67            75   
26               JOSEPH N.         CAPPELLA             40            49   
27             KRYSTYNA B.          WARCHOL              8            13   
28                  ALBERT             HUNT             26            27   
29               DWIGHT D.       EISENHOWER             23            25   
..                     ...              ...            ...           ...   
730              JOSEPH N.         CAPPELLA             30            30   
731                LIK SAM             CHAN             21            21   
732          DWAYNE TAYLOR            BOOTH             39            39   
733          MATTHEW BROOK        O'DONNELL             22            22   
734               AARON M.          SHAPIRO             12            12   
735              CAROLYN A           MARVIN             20            21   
736                 ALBERT             HUNT             27            27   
737          RAYYA SUNAYMA          EL ZEIN             20            20   
738             DANIEL RAY         GRINBERG             21            21   
739              CARLIN P.           ROMANO             14            14   
740                  EYTAN           GILBOA             22            22   
741              DWIGHT D.       EISENHOWER             35            35   
742            SCOTT DAVID            REICH             35            35   
743                KENNETH           WINNEG             21            21   
744                 SANDRA  GONZALEZ-BAILON             21            21   
745                YPHTACH           LELKES             21            21   
746               DAMON M.          CENTOLA             14            14   
747               SUSAN D.             HAAS              4             5   
748                DAVID J           DUTWIN             14            15   
749                 MURALI           BALAJI              3             3   
750                 GUOBIN             YANG             58

If we want to be able to do calculations and statistical analyses with the data, all the columns must be categorized as numeric and not string values.

In [13]:
# converting to numeric
comm_revs_df2[['CourseQuality', 'InstructorQuality', 'Difficulty',
       'AmountLearned', 'WorkRequired', 'StimulateInterest',
       'InstructorAccess', 'CommAbility', 'ReadingsValue', 'TAQuality',
       'RecommendMajor', 'RecommendNonMajor', 'num_reviewers',
            'num_students']] = comm_revs_df2[['CourseQuality', 'InstructorQuality', 'Difficulty',
       'AmountLearned', 'WorkRequired', 'StimulateInterest',
       'InstructorAccess', 'CommAbility', 'ReadingsValue', 'TAQuality',
       'RecommendMajor', 'RecommendNonMajor', 'num_reviewers', 'num_students']].apply(pd.to_numeric)

The actual course number (e.g. 123) is embedded in the CourseCode, but we should extract it to compare classes:

In [14]:
# splitting the course code column to extract just 'level' (course code)
comm_revs_df2['level']= comm_revs_df2['CourseCode'].apply(lambda v: v.split("-")[1])
comm_revs_df2

instructor_first  instructor_last  num_reviewers  num_students  \
0                  JESSICA          FISHMAN             83            89   
1                  VINCENT            PRICE            133           181   
2               MARIAELENA       BARTESAGHI             23            26   
3                  KATHRYN          KOLBERT             73           112   
4                   AMY B.           JORDAN             75            89   
5                    EMILY             WEST             23            24   
6                JOSEPHINE         FERRIGNO             18            25   
7                  RONA J.        BUCHALTER             13            14   
8                CAROLYN A           MARVIN             17            18   
9          MARGARET LOUISE        WOODSTOCK             17            18   
10               JOSEPH G.            TUROW             56            92   
11            OSCAR HOWARD            GANDY              5             6   
12               JOSEPH N.         CAPPELLA             48            56   
13                   KLAUS     KRIPPENDORFF             15            19   
14               DWIGHT D.       EISENHOWER             22            23   
15                    JOHN          JEMMOTT              6             6   
16                 MICHAEL            HAGEN             12            16   
17                   ELIHU             KATZ              5            21   
18                  ALBERT             HUNT             14            31   
19                 RONA J.        BUCHALTER             24            26   
20               JOSEPH G.            TUROW             94           192   
21                CARRIE D           RICKEY             49            79   
22              DEBORAH L.       LINEBARGER             26            37   
23              KIMBERLY D            WOOLF             21            24   
24                  CLAIRE           WARDLE             18            20   
25               AMY SARCH         SCHOPICK             67            75   
26               JOSEPH N.         CAPPELLA             40            49   
27             KRYSTYNA B.          WARCHOL              8            13   
28                  ALBERT             HUNT             26            27   
29               DWIGHT D.       EISENHOWER             23            25   
..                     ...              ...            ...           ...   
730              JOSEPH N.         CAPPELLA             30            30   
731                LIK SAM             CHAN             21            21   
732          DWAYNE TAYLOR            BOOTH             39            39   
733          MATTHEW BROOK        O'DONNELL             22            22   
734               AARON M.          SHAPIRO             12            12   
735              CAROLYN A           MARVIN             20            21   
736                 ALBERT             HUNT             27            27   
737          RAYYA SUNAYMA          EL ZEIN             20            20   
738             DANIEL RAY         GRINBERG             21            21   
739              CARLIN P.           ROMANO             14            14   
740                  EYTAN           GILBOA             22            22   
741              DWIGHT D.       EISENHOWER             35            35   
742            SCOTT DAVID            REICH             35            35   
743                KENNETH           WINNEG             21            21   
744                 SANDRA  GONZALEZ-BAILON             21            21   
745                YPHTACH           LELKES             21            21   
746               DAMON M.          CENTOLA             14            14   
747               SUSAN D.             HAAS              4             5   
748                DAVID J           DUTWIN             14            15   
749                 MURALI           BALAJI              3             3   
750                 GUOBIN             YANG             58

Re-ordering the columns to emphasize the information we primarily need:

In [15]:
# re-ordering the columns

comm_revs_df2 = comm_revs_df2[['CourseCode','level','Section','CourseTitle','Semester','instructor_first',
 'instructor_last','CourseQuality','InstructorQuality','Difficulty','AmountLearned','WorkRequired',
'StimulateInterest','InstructorAccess','CommAbility','ReadingsValue','TAQuality','RecommendMajor',
'RecommendNonMajor','num_reviewers','num_students']]

comm_revs_df2.head()

CourseCode level Section               CourseTitle Semester  \
0  COMM-123-001   123     001    COMM & POPULAR CULTURE    2002A   
1  COMM-125-001   125     001    COMMUNICATION BEHAVIOR    2002A   
2  COMM-125-601   125     601    COMMUNICATION BEHAVIOR    2002A   
3  COMM-175-401   175     401  ARGUMENT & PUBL ADVOCACY    2002A   
4  COMM-225-001   225     001          CHILDREN & MEDIA    2002A   

  instructor_first instructor_last  CourseQuality  InstructorQuality  \
0          JESSICA         FISHMAN           2.43               2.88   
1          VINCENT           PRICE           2.11               2.62   
2       MARIAELENA      BARTESAGHI           2.74               3.17   
3          KATHRYN         KOLBERT           1.89               2.26   
4           AMY B.          JORDAN           3.20               3.49   

   Difficulty  ...  WorkRequired  StimulateInterest  InstructorAccess  \
0        2.28  ...          2.28               2.81              2.84   
1        2.27  ...          2.32               1.92              2.62   
2        2.13  ...          2.13               3.30              3.61   
3        1.49  ...          1.83               1.79              2.34   
4        2.13  ...          2.29               3.07              3.24   

   CommAbility  ReadingsValue  TAQuality  RecommendMajor  RecommendNonMajor  \
0         2.88           2.24        NaN            3.21               2.53   
1         2.84           2.14        NaN            3.15               1.47   
2         3.17           3.17        NaN            3.17               2.09   
3         2.37           1.99        NaN            2.42               1.86   
4         3.49           2.49        NaN            3.60               2.87   

   num_reviewers  num_students  
0             83            89  
1            133           181  
2             23            26  
3             73           112  
4             75            89  

[5 rows x 21 columns]

#### Exploring the data

In [16]:
comm_revs_df2.shape

(760, 21)

#### There are 760 rows (courses) here, and 21 remaining columns (variables). How can we view a list of the courses?

In [17]:
comm_courses = sorted(comm_revs_df2['CourseCode'].unique())
comm_courses

['AFRC-287-401',
 'AFRC-306-401',
 'AFRC-387-401',
 'ANTH-123-401',
 'ANTH-141-401',
 'ANTH-459-401',
 'ASAM-201-401',
 'ASAM-201-601',
 'CINE-150-401',
 'CINE-296-401',
 'CINE-296-402',
 'COMM-105-401',
 'COMM-108-301',
 'COMM-112-301',
 'COMM-123-001',
 'COMM-123-900',
 'COMM-123-910',
 'COMM-123-920',
 'COMM-125-001',
 'COMM-125-601',
 'COMM-125-910',
 'COMM-125-920',
 'COMM-130-001',
 'COMM-130-910',
 'COMM-130-920',
 'COMM-140-001',
 'COMM-140-401',
 'COMM-140-900',
 'COMM-140-910',
 'COMM-175-401',
 'COMM-205-301',
 'COMM-206-401',
 'COMM-207-401',
 'COMM-208-301',
 'COMM-209-301',
 'COMM-210-001',
 'COMM-210-301',
 'COMM-211-301',
 'COMM-213-301',
 'COMM-217-301',
 'COMM-218-910',
 'COMM-221-001',
 'COMM-224-001',
 'COMM-224-301',
 'COMM-225-001',
 'COMM-225-601',
 'COMM-225-910',
 'COMM-225-920',
 'COMM-226-001',
 'COMM-226-401',
 'COMM-226-601',
 'COMM-226-670',
 'COMM-226-920',
 'COMM-230-001',
 'COMM-237-001',
 'COMM-237-301',
 'COMM-237-900',
 'COMM-237-910',
 'COMM-238-001

#### Note there are some cross-listed courses here listed under different departments. Next, can we get an alphabetical list of instructors?

In [18]:
sorted(comm_revs_df2['instructor_last'].unique())

['AGHA',
 'AL-MARASHI',
 'ALBARRACIN',
 'ALLEN',
 'ARSENAULT',
 'BALAJI',
 'BARMADA',
 'BARRETT',
 'BARTESAGHI',
 'BARUH',
 'BELL',
 'BEN-PORATH',
 'BERGER',
 'BERGERE',
 'BERMAN',
 'BEUTIN',
 'BIGMAN-GALIMORE',
 'BLEAKLEY',
 'BLONDHEIM',
 'BLUMENTHAL',
 'BOCK',
 'BOOTH',
 'BRECHMAN',
 'BRIDGES',
 'BROAD',
 'BRUNEAU',
 'BRUTGER',
 'BUCHALTER',
 'BUMGARNER',
 'CAPPELLA',
 'CARLSON',
 'CENTOLA',
 'CHAN',
 'CHATMAN',
 'CHEN',
 'CHERNIN',
 'CHESTER',
 'CHOMSKY',
 'CHRISTENSEN',
 'COPPA',
 'COYER',
 'CROCCO',
 'DANILLER',
 'DECHERNEY',
 'DELLI CARPINI',
 'DILLIPLANE',
 'DUFFY',
 'DUTWIN',
 'EISENHOWER',
 'EL ZEIN',
 'ERDENER',
 'ERIGHA',
 'FALK',
 'FALZONE',
 'FARRALL',
 'FELDMAN',
 'FELSENTHAL',
 'FELZENBERG',
 'FERRARI',
 'FERRIGNO',
 'FINEMAN',
 'FISHMAN',
 'FRAZE',
 'FRERES',
 'GABBADON',
 'GANDY',
 'GARFIELD',
 'GARRY',
 'GIBSON',
 'GILBOA',
 'GILLIAM',
 'GIRGINOVA',
 'GLANVILLE',
 'GOLDMAN',
 'GOLDTHWAITE YOUNG',
 'GONZALEZ-BAILON',
 'GOTTFRIED',
 'GRIEBLING',
 'GRINBERG',
 'GROSS',
 

#### Generating some initial summary statistics on the dataset:

In [19]:
comm_revs_df2.describe()

CourseQuality  InstructorQuality  Difficulty  AmountLearned  \
count     760.000000         760.000000  757.000000     701.000000   
mean        2.969829           3.225632    2.367649       2.955478   
std         0.578296           0.543973    0.474776       0.499488   
min         0.790000           0.620000    1.000000       1.110000   
25%         2.580000           2.860000    2.040000       2.640000   
50%         3.000000           3.330000    2.380000       3.000000   
75%         3.400000           3.670000    2.670000       3.310000   
max         4.000000           4.000000    4.000000       4.000000   

       WorkRequired  StimulateInterest  InstructorAccess  CommAbility  \
count    757.000000         757.000000        757.000000   757.000000   
mean       2.505931           2.988322          3.155033     3.200898   
std        0.497504           0.680073          0.554780     0.528837   
min        1.170000           0.430000          0.950000     0.600000   
25%        2.180000           2.500000          2.830000     2.880000   
50%        2.470000           3.090000          3.220000     3.290000   
75%        2.790000           3.550000          3.560000     3.610000   
max        4.000000           4.000000          4.000000     4.000000   

       ReadingsValue   TAQuality  RecommendMajor  RecommendNonMajor  \
count     697.000000  266.000000      701.000000         700.000000   
mean        2.622281    2.966316        3.246091           2.377171   
std         0.504454    0.651513        0.518802           0.802207   
min         0.670000    0.000000        0.850000           0.000000   
25%         2.290000    2.572500        2.940000           1.830000   
50%         2.630000    3.000000        3.320000           2.470000   
75%         2.930000    3.350000        3.670000           3.000000   
max         4.000000    4.000000        4.000000           4.000000   

       num_reviewers  num_students  
count     760.000000    760.000000  
mean       29.692105     34.132895  
std        29.269093     36.217059  
min         1.000000      1.000000  
25%        13.000000     15.000000  
50%        20.000000     21.000000  
75%        32.000000     35.000000  
max       158.000000    194.000000

#### We can see some initial insights from looking at the above:
* Not every variable has a count of 760, although there were 760 courses here. Not every course was reviewed along every dimension.
* The mean values are pretty close to median values, suggesting there may not be such skewed distribution in many of the data.
* The maximum is 4.00 on the Penn Course Review scale. In general the analyses to come are working within a pretty small range (with minimum reviews at 0 and maximum at 4.00 - so effect sizes are going to be small. Something to keep in mind.
* Also import to note that the reviews begin in 2002, which is not the totality of historical data but is relevant to modern course analysis.

#### What if we just wanted the means for each category?

In [20]:
comm_revs_df2.mean()

CourseQuality         2.969829
InstructorQuality     3.225632
Difficulty            2.367649
AmountLearned         2.955478
WorkRequired          2.505931
StimulateInterest     2.988322
InstructorAccess      3.155033
CommAbility           3.200898
ReadingsValue         2.622281
TAQuality             2.966316
RecommendMajor        3.246091
RecommendNonMajor     2.377171
num_reviewers        29.692105
num_students         34.132895
dtype: float64

#### And how did these means change over time?

In [21]:
comm_revs_df2.groupby('Semester').mean()

CourseQuality  InstructorQuality  Difficulty  AmountLearned  \
Semester                                                                
2002A          2.803684           3.074211    2.326842       2.737368   
2002C          2.824118           3.115882    2.357059       2.794706   
2003A          2.802143           3.070714    2.395000       2.825714   
2003B          3.195000           3.195000    2.395000       3.150000   
2003C          2.871765           3.228824    2.219412       2.829412   
2004A          3.001429           3.232857    2.417857       2.920000   
2004C          3.067059           3.330588    2.402353       2.999412   
2005A          2.908125           3.189375    2.492500       2.828750   
2005B          3.015714           3.288571    2.445714       3.162857   
2005C          2.900588           3.204118    2.459412       2.818824   
2006A          2.980000           3.265000    2.431250       2.895625   
2006B          2.950000           3.228000    2.394000       2.892000   
2006C          3.024375           3.337500    2.488125       2.948750   
2007A          2.979444           3.268333    2.327222       2.883889   
2007B          3.311667           3.531667    2.305000       3.431667   
2007C          2.879474           3.182105    2.359474       2.786316   
2008A          3.085500           3.307000    2.530500       3.070500   
2008B          2.988750           3.235000    2.577500       3.061250   
2008C          3.083636           3.340455    2.426818       3.071364   
2009A          2.929200           3.146800    2.558400       2.889600   
2009B          2.822500           3.161250    2.326250       2.698750   
2009C          2.846190           3.126667    2.445238       2.754762   
2010A          2.875882           3.110000    2.304118       2.821765   
2010B          3.001250           3.125000    2.382500       2.936250   
2010C          2.985238           3.250952    2.402381       2.954762   
2011A          2.866316           3.010000    2.381579       2.786842   
2011B          3.054444           3.230000    2.047778       3.095556   
2011C          3.004444           3.305556    2.533333       2.952778   
2012A          3.028095           3.250000    2.361429       3.000000   
2012B          3.242500           3.612500    1.897500       3.140000   
2012C          3.148571           3.399524    2.317222       3.065000   
2013A          3.003103           3.131034    2.306552       3.047222   
2013C          2.848947           3.203158    2.298947       2.875000   
2014A          3.020500           3.265500    2.366500       3.052353   
2014C          2.847000           3.100500    2.500500       2.873333   
2015A          2.991852           3.221481    2.507037       3.143182   
2015C          2.898000           3.184000    2.521333       2.983333   
2016A          2.862273           3.131364    2.305909       3.026667   
2016C          3.079545           3.295000    2.250000       3.141667   
2017A          2.854500           3.086000    2.250500       3.008824   
2017C          3.061739           3.401739    2.262609       3.143636   
2018A          2.907143           3.193810    2.105238       3.031579   
2018B          3.442000           3.774000    2.042000            NaN   
2018C          2.900455           3.146818    2.391364       2.963000   
2019A          3.056250           3.277500    2.310833       3.088095   
2019B          3.488889           3.684444    2.074444            NaN   

          WorkRequired  StimulateInterest  InstructorAccess  CommAbility  \
Semester                                                                   
2002A         2.437895           2.720526          3.169474     3.086316   
2002C         2.570000           2.728235          3.106471     3.053529   
2003A         2.452857           2.810000          2.958571     3.119286   
2003B         2.510000           3.045000          3.435000     3.070000   
2003C         2.384118           2.898235        

#### Summer courses may have a different vibe altogether. We can do the above analysis procedure without summer courses.

We will have to make a non-summer courses dataframe by looking for the letters A (for spring) and C (for fall) in the semesters column.

In [22]:
# Create a subset of the data that does not include summer courses
# look for the letter A (spring) or C (fall) 
#at the end of the 'Semester' string

comm_nosummer = comm_revs_df2[comm_revs_df2['Semester'].str.contains("A|C")]

comm_nosummer.groupby('Semester').mean()

CourseQuality  InstructorQuality  Difficulty  AmountLearned  \
Semester                                                                
2002A          2.803684           3.074211    2.326842       2.737368   
2002C          2.824118           3.115882    2.357059       2.794706   
2003A          2.802143           3.070714    2.395000       2.825714   
2003C          2.871765           3.228824    2.219412       2.829412   
2004A          3.001429           3.232857    2.417857       2.920000   
2004C          3.067059           3.330588    2.402353       2.999412   
2005A          2.908125           3.189375    2.492500       2.828750   
2005C          2.900588           3.204118    2.459412       2.818824   
2006A          2.980000           3.265000    2.431250       2.895625   
2006C          3.024375           3.337500    2.488125       2.948750   
2007A          2.979444           3.268333    2.327222       2.883889   
2007C          2.879474           3.182105    2.359474       2.786316   
2008A          3.085500           3.307000    2.530500       3.070500   
2008C          3.083636           3.340455    2.426818       3.071364   
2009A          2.929200           3.146800    2.558400       2.889600   
2009C          2.846190           3.126667    2.445238       2.754762   
2010A          2.875882           3.110000    2.304118       2.821765   
2010C          2.985238           3.250952    2.402381       2.954762   
2011A          2.866316           3.010000    2.381579       2.786842   
2011C          3.004444           3.305556    2.533333       2.952778   
2012A          3.028095           3.250000    2.361429       3.000000   
2012C          3.148571           3.399524    2.317222       3.065000   
2013A          3.003103           3.131034    2.306552       3.047222   
2013C          2.848947           3.203158    2.298947       2.875000   
2014A          3.020500           3.265500    2.366500       3.052353   
2014C          2.847000           3.100500    2.500500       2.873333   
2015A          2.991852           3.221481    2.507037       3.143182   
2015C          2.898000           3.184000    2.521333       2.983333   
2016A          2.862273           3.131364    2.305909       3.026667   
2016C          3.079545           3.295000    2.250000       3.141667   
2017A          2.854500           3.086000    2.250500       3.008824   
2017C          3.061739           3.401739    2.262609       3.143636   
2018A          2.907143           3.193810    2.105238       3.031579   
2018C          2.900455           3.146818    2.391364       2.963000   
2019A          3.056250           3.277500    2.310833       3.088095   

          WorkRequired  StimulateInterest  InstructorAccess  CommAbility  \
Semester                                                                   
2002A         2.437895           2.720526          3.169474     3.086316   
2002C         2.570000           2.728235          3.106471     3.053529   
2003A         2.452857           2.810000          2.958571     3.119286   
2003C         2.384118           2.898235          3.141765     3.134118   
2004A         2.470000           2.945000          3.007857     3.241429   
2004C         2.517059           2.982941          3.166471     3.295882   
2005A         2.551250           2.942500          3.044375     3.171875   
2005C         2.475294           2.881176          3.189412     3.126471   
2006A         2.556875           2.917500          3.130000     3.256250   
2006C         2.640000           3.067500          3.225625     3.313125   
2007A         2.486111           2.968889          3.306667     3.223889   
2007C         2.495263           2.890526          3.191053     3.134737   
2008A         2.642500           3.041000          3.307500     3.340000   
2008C         2.448636           3.099091          3.263636     3.247273   
2009A         2.848400           2.921200          3.122800     3.110800   
2009C         2.

#### What correlations can we see between reviews in different categories?

In [23]:
comm_nosummer.corr()

CourseQuality  InstructorQuality  Difficulty  \
CourseQuality           1.000000           0.919831    0.092403   
InstructorQuality       0.919831           1.000000    0.064127   
Difficulty              0.092403           0.064127    1.000000   
AmountLearned           0.899660           0.821691    0.222857   
WorkRequired            0.108361           0.105471    0.779739   
StimulateInterest       0.912868           0.919087    0.061489   
InstructorAccess        0.546393           0.632354   -0.042500   
CommAbility             0.889546           0.929821    0.071461   
ReadingsValue           0.684639           0.595534    0.153255   
TAQuality               0.348113           0.313895   -0.003397   
RecommendMajor          0.868370           0.828410    0.014992   
RecommendNonMajor       0.709451           0.633118   -0.145721   
num_reviewers          -0.159587          -0.155523    0.034093   
num_students           -0.170281          -0.165039    0.038213   

                   AmountLearned  WorkRequired  StimulateInterest  \
CourseQuality           0.899660      0.108361           0.912868   
InstructorQuality       0.821691      0.105471           0.919087   
Difficulty              0.222857      0.779739           0.061489   
AmountLearned           1.000000      0.216260           0.829168   
WorkRequired            0.216260      1.000000           0.076057   
StimulateInterest       0.829168      0.076057           1.000000   
InstructorAccess        0.499151      0.112877           0.521308   
CommAbility             0.820380      0.096095           0.905718   
ReadingsValue           0.726558      0.168426           0.628297   
TAQuality               0.276898      0.147535           0.291590   
RecommendMajor          0.794125      0.006789           0.831403   
RecommendNonMajor       0.609065     -0.252627           0.703331   
num_reviewers          -0.165588     -0.189109          -0.089101   
num_students           -0.176756     -0.180321          -0.116720   

                   InstructorAccess  CommAbility  ReadingsValue  TAQuality  \
CourseQuality              0.546393     0.889546       0.684639   0.348113   
InstructorQuality          0.632354     0.929821       0.595534   0.313895   
Difficulty                -0.042500     0.071461       0.153255  -0.003397   
AmountLearned              0.499151     0.820380       0.726558   0.276898   
WorkRequired               0.112877     0.096095       0.168426   0.147535   
StimulateInterest          0.521308     0.905718       0.628297   0.291590   
InstructorAccess           1.000000     0.586202       0.421563   0.305334   
CommAbility                0.586202     1.000000       0.641001   0.349735   
ReadingsValue              0.421563     0.641001       1.000000   0.293327   
TAQuality                  0.305334     0.349735       0.293327   1.000000   
RecommendMajor             0.508155     0.822837       0.564114   0.290244   
RecommendNonMajor          0.315645     0.651621       0.491229   0.145523   
num_reviewers             -0.340277    -0.100136      -0.185204  -0.157227   
num_students              -0.340139    -0.108248      -0.188079  -0.165168   

                   RecommendMajor  RecommendNonMajor  num_reviewers  \
CourseQuality            0.868370           0.709451      -0.159587   
InstructorQuality        0.828410           0.633118      -0.155523   
Difficulty               0.014992          -0.145721       0.034093   
AmountLearned            0.794125           0.609065      -0.165588   
WorkRequired             0.006789          -0.252627      -0.189109   
StimulateInterest        0.831403           0.703331      -0.089101   
InstructorAccess         0.508155           0.315645      -0.340277   
CommAbility              0.822837           0.651621      -0.100136   
ReadingsValue            0.564114           0.491229      -0.185204   
TAQuality                0.290244           0.145523      -0.157227   
RecommendMajor           

#### Initial insights from above:
* Predictably, course quality and instructor quality are very tightly correlated (.92)
* Difficulty is not very highly correlated with course quality or instructor quality at all. COMM students are capable of distinguishing between a 'good' class and an easy one.
* Value of Readings correlates pretty highly with Amount Learned, suggesting readings are an important part of a course.
* As Number of Students increases, Instructor Access (predictably) falls, and there are small negative correlations with Course Quality and Instructor Quality. 


#### How many students enroll in COMM classes each semester?

Grouping by semester and giving the number of students for each:

In [24]:
comm_numstudents = comm_nosummer.groupby('Semester')[['num_students']].sum()
comm_numstudents.reset_index()

Semester  num_students
0     2002A           866
1     2002C           659
2     2003A           618
3     2003C           735
4     2004A           800
5     2004C           799
6     2005A           668
7     2005C           857
8     2006A           761
9     2006C           840
10    2007A           753
11    2007C           823
12    2008A           690
13    2008C          1066
14    2009A           672
15    2009C           714
16    2010A           608
17    2010C           726
18    2011A           590
19    2011C           585
20    2012A           547
21    2012C           640
22    2013A           897
23    2013C           827
24    2014A           591
25    2014C           698
26    2015A           635
27    2015C           616
28    2016A           865
29    2016C           629
30    2017A           672
31    2017C           635
32    2018A           677
33    2018C           585
34    2019A           670

#### How many COMM courses has each instructor taught?

In [25]:
comm_revs_df2.groupby('instructor_last').size().sort_values(ascending=False)

instructor_last
EISENHOWER           44
JORDAN               42
TUROW                25
MESSARIS             25
PAXTON               21
MARVIN               21
ROMANO               21
FELZENBERG           20
HUNT                 18
JAMIESON             17
BLEAKLEY             17
CAPPELLA             14
PEARL                13
HAAS                 12
JACKSON              12
THEOPHANO            12
LINEBARGER           10
KATZ                  9
SENDER                9
PRICE                 9
ZELIZER               9
YANG                  9
GARRY                 9
WOOLF                 9
WINNEG                9
HART                  9
HAMPTON               8
DUTWIN                8
BOOTH                 8
MOEHLER               7
                     ..
SHAW                  1
HENRICHSEN            1
THORSON               1
MAURANTONIO           1
RICKEY                1
FISHMAN               1
FRAZE                 1
FRERES                1
GABBADON              1
WALDMAN               1


#### How many times has each COMM course been offered?

In [26]:
comm_revs_df2['level'].value_counts()

395    35
125    32
130    31
225    28
123    26
299    25
398    25
275    19
226    19
323    17
290    16
322    16
262    15
300    15
491    14
339    12
340    12
374    11
397    11
237    10
495    10
494     9
230     9
399     9
390     9
210     9
377     8
140     8
413     8
330     7
       ..
349     1
175     1
302     1
206     1
445     1
205     1
421     1
315     1
278     1
191     1
287     1
348     1
326     1
409     1
294     1
381     1
260     1
531     1
418     1
105     1
353     1
213     1
368     1
561     1
420     1
337     1
241     1
382     1
307     1
352     1
Name: level, Length: 161, dtype: int64

#### Which level of COMM classes is offered the most often?

In [27]:
# creating a new column to determine what hundred level each course is 
comm_revs_df2['hundred'] = comm_revs_df2.level.str[:1]

In [28]:
# within those what does the breakdown look like?
by_level = comm_revs_df2.groupby('hundred')
by_level.size()

hundred
0      3
1    112
2    207
3    301
4    135
5      2
dtype: int64

In [29]:
# within those levels which courses are offered how many times?
counts_by_level = by_level['level'].value_counts()
counts_by_level

hundred  level
0        024       3
1        125      32
         130      31
         123      26
         140       8
         141       7
         150       2
         105       1
         108       1
         110       1
         112       1
         175       1
         191       1
2        225      28
         299      25
         226      19
         275      19
         290      16
         262      15
         237      10
         210       9
         230       9
         238       6
         292       5
         245       4
         211       3
         224       3
         281       3
         282       3
         296       3
                  ..
4        408       3
         415       3
         416       3
         481       3
         496       3
         406       2
         407       2
         411       2
         419       2
         433       2
         441       2
         454       2
         459       2
         485       2
         409       1
         412       

We have finally arrived at a solid understanding of the cleaned data:

In [30]:
comm_revs_df2

CourseCode level Section  \
0    COMM-123-001   123     001   
1    COMM-125-001   125     001   
2    COMM-125-601   125     601   
3    COMM-175-401   175     401   
4    COMM-225-001   225     001   
5    COMM-262-601   262     601   
6    COMM-280-601   280     601   
7    COMM-299-001   299     001   
8    COMM-322-301   322     301   
9    COMM-322-601   322     601   
10   COMM-330-001   330     001   
11   COMM-334-301   334     301   
12   COMM-350-001   350     001   
13   COMM-360-301   360     301   
14   COMM-395-301   395     301   
15   COMM-398-302   398     302   
16   COMM-398-303   398     303   
17   COMM-496-301   496     301   
18   COMM-498-301   498     301   
19   COMM-125-601   125     601   
20   COMM-130-001   130     001   
21   COMM-140-001   140     001   
22   COMM-225-001   225     001   
23   COMM-225-601   225     601   
24   COMM-226-601   226     601   
25   COMM-238-001   238     001   
26   COMM-275-001   275     001   
27   COMM-299-001   299     001   
28   COMM-323-301   323     301   
29   COMM-395-301   395     301   
..            ...   ...     ...   
730  COMM-275-001   275     001   
731  COMM-290-301   290     301   
732  COMM-292-001   292     001   
733  COMM-313-301   313     301   
734  COMM-317-301   317     301   
735  COMM-322-301   322     301   
736  COMM-323-301   323     301   
737  COMM-352-301   352     301   
738  COMM-353-301   353     301   
739  COMM-377-301   377     301   
740  COMM-390-301   390     301   
741  COMM-395-301   395     301   
742  COMM-395-301   395     301   
743  COMM-397-301   397     301   
744  COMM-407-301   407     301   
745  COMM-441-301   441     301   
746  COMM-459-301   459     301   
747  COMM-491-301   491     301   
748  PSCI-332-401   332     401   
749  SAST-110-401   110     401   
750  SOCI-238-401   238     401   
751  COMM-125-910   125     910   
752  COMM-125-910   125     910   
753  COMM-130-920   130     920   
754  COMM-130-920   130     920   
755  COMM-225-910   225     910   
756  COMM-290-900   290     900   
757  COMM-290-910   290     910   
758  COMM-290-910   290     910   
759  COMM-339-920   339     920   

                                           CourseTitle Semester  \
0                               COMM & POPULAR CULTURE    2002A   
1                               COMMUNICATION BEHAVIOR    2002A   
2                               COMMUNICATION BEHAVIOR    2002A   
3                             ARGUMENT & PUBL ADVOCACY    2002A   
4                                     CHILDREN & MEDIA    2002A   
5                                 VISUAL COMMUNICATION    2002A   
6                                 COMM & GLOBALIZATION    2002A   
7                             COMMUNICATIONS INTERNSHP    2002A   
8                                FREEDOM OF EXPRESSION    2002A   
9                                FREEDOM OF EXPRESSION    2002A   
10                             STRUC/FUNCT ADVERTISING    2002A   
11                                 THE INFORMATION AGE    2002A   
12                                     NON-VERBAL COMM    2002A   
13                             LANG SOC CONSTRUCT REAL    2002A   
14                               COMM & THE PRESIDENCY    2002A   
15   SP TOPICS IN MASS COMM: Seminar in Health Comm...    2002A   
16                              ELECTIONS & MASS MEDIA    2002A   
17                             LEISURE, COMM & CULTURE    2002A   
18                               CONTMP POL PLCY PRESS    2002A   
19                              COMMUNICATION BEHAVIOR    2003A   
20                                    MASS MEDIA & SOC    2003A   
21                             INTRO FILM,FORM,CONTEXT    2003A   
22                                    CHILDREN & MEDIA    2003A   
23                                    CHILDREN & MEDIA    2003A   
24                             INTRO TO POLITICAL COMM    2003A   
25                            HIST OF GENDER IN ADVERT    2003A   
26                         

In [31]:
# saving this df to the cleaned data folder
comm_revs_df2.to_csv('../data/clean_data/comm_revs_df2',index=False)